File: 2015/25/python/main.py
import re
from pathlib import Path

def get_code(r: int, c: int) -> int:
    current_value = 20151125
    value = lambda x: (x * 252533) % 33554393

    # Calculate the target position in the sequence
    target_index = sum(range(1, r + c)) - (r - 1)

    # Generate values until the target index
    for _ in range(1, target_index):
        current_value = value(current_value)

    return current_value

def main() -> None:
    rows, cols = [int(x) for x in re.findall(r"\d+", Path("../input.txt").read_text().strip())]
    code = get_code(rows, cols)
    print(f"The code to start the machine is: {code}")

if __name__ == "__main__":
    main()

File: 2015/25/rust/src/main.rs
use regex::Regex;
use std::fs;

fn calc_value(x: u64) -> u64 {
    (x * 252533) % 33554393
}

fn get_code(r: u64, c: u64) -> u64 {
    let mut current_value: u64 = 20151125;
    let target_index: u64 = (1..(r + c)).sum::<u64>() - (r - 1);

    for _ in 1..target_index {
        current_value = calc_value(current_value);
    }
    current_value
}

fn main() {
    let content = fs::read_to_string("../input.txt")
        .expect("Failed to read input file")
        .trim()
        .to_string();

    let re = Regex::new(r"\d+").unwrap();
    let numbers: Vec<u64> = re
        .find_iter(&content)
        .map(|m| m.as_str().parse::<u64>().unwrap())
        .collect();
    let code = get_code(numbers[0], numbers[1]);
    println!("The code to start the machine is: {code}");
}
